frontends/p4/frontend.cpp

```diff
@@ -148,6 +149,9 @@ const IR::P4Program *FrontEnd::run(const CompilerOptions &options, const IR::P4P
         new TypeInference(&refMap, &typeMap, false),  // insert casts
         new ValidateMatchAnnotations(&typeMap),
         new DefaultArguments(&refMap, &typeMap),  // add default argument values to parameters
+#ifdef NDEBUG
```

Comment:
First of all, this NDEBUG is for the compiler source code. You should remove assert/assume only if the p4 compiler is invoked with NDEBUG, not gcc.
Second, this is too early to remove the calls. If the calls have side-effects they should still happen. So the removal should happen after the SideEffects pass.
Third, I am not sure we actually want this part of the front-end; perhaps the mid-end.

Reply:
If I understood you well, p4 compiler should remove assert/assume if it is invoked with  -DNDEBUG (and default) option and leave it if it is invoked with -UNDEBUG. I moved this pass to be the first in the mid-end, before any of the optimizations is done. Feel free to correct me if I misunderstood it.

Reply:
What Mihai means is that when you write `#ifdef NDEBUG`, you are using a preprocessor flag that is meant for the C++ compiler with which you compile p4c. The `assert` / `assume` statements should be removed if the *p4c* compiler is invoked with `-DNDEBUG`. When I compile a C / C++ program, I choose whether asserts are removed by providing `-DNEBUG` or not. Whether the C / C++ compiler was itself compiled with `-DNEBUG` is completely irrelevant. The p4c invocation `p4c -DNEBUG <path to P4 program>` should remove the `assert` / `assume` statements. I would argue that using a preprocessor flag for this is probably not ideal. What about `p4c -ndebug <path to P4 program>` instead?